### ISSUE TITLE
BOSH DNS should be reachable by workers

### ISSUE THREAD

**Issue by sharms** [concourse/concourse#2482]
# Feature Request
As a user of CloudFoundry and BOSH, our systems are configured to use BOSH DNS.  This DNS cannot currently be resolved by default with Concourse workers without adding `iptables` rules by various out-of-band and hard to maintain methods.

## What challenge are you facing?

By default Concourse should allow containers to reach `169.254.0.2` to conform with the default BOSH configuration

## A Modest Proposal

Add iptables rules by default which allow workers to contact BOSH DNS

**Comment by MattSurabian** [comment]
After rolling out `bosh-dns` we saw our pipeline run times skyrocket which led us here. Now we know that workers trying to fetch resources hang until a timeout for `169.254.0.2` forces them to the next resolver in `/etc/resolv.conf`. 

Hoping this wouldn't be an issue re: worker principles; my understanding is that `bosh-dns` will soon be the default which would make this a bit more pressing. I only wish I had a decent suggestion on how to make a generic-ish configuration property to account for cases like this where "external" or otherwise sequestered resources need to be reachable from the worker's containers.

**Comment by tlwr** [comment]
While upgrading cf-deployment from pre-v3.0.0 to v3.6.0 we followed the CF guidance to move bosh-dns from cf-deployment to our global bosh runtime-config.

### Behaviour

We observed the same issues:
- able to do DNS resolution using bosh-dns on the host
- able to do DNS resolution using VPC cidr inside containers
- not able to do DNS resolution using bosh-dns inside the containers

---

### Verification

Ruby to illustrate (run via `fly hijack`):
```
require 'resolv'
require 'benchmark'

[{
  name: 'backup',
  resolver: Resolv.new([Resolv::DNS.new(nameserver:'10.0.0.2')]),
},{
  name: 'default',
  resolver: Resolv::DefaultResolver,
},{
  name: 'bosh',
  resolver: Resolv.new([Resolv::DNS.new(nameserver:'169.254.0.2')]),
}]
.map { |r| r.values_at(:name, :resolver) }
.map do |name, resolver|
  print "resolver #{name.ljust(10, ' ')} => "
  print (Benchmark.realtime do
    resolver.getaddress 'login.jonm-dep.dev.cloudapps.digital'
  end)
  puts ' seconds'
end
```
```
resolver backup     => 0.005594876998657128 seconds
resolver default    => 5.004335433997767 seconds
resolver bosh       => Traceback (most recent call last): ...
```

i.e. the DNS resolver would try bosh-dns, timeout after 5s, then use the VPC DNS, this caused many jobs in our concourse pipelines to tcp dial timeout.

---

### Workaround

We have our own concourse manifest and we manually overrode the default `dns_servers` value to point to our VPC DNS:
```
instance_groups:
  - name: concourse
    instances: 1
    jobs:
      - name: garden
        release: garden-runc
        properties:
          garden:
            # Override concourse containers to use VPC DNS instead of bosh-dns
            # See https://github.com/concourse/concourse/issues/2482 for details
            dns_servers:
              - 10.0.0.2
```

### Collaborators
- @richardtowers
- @jpluscplusm

**Comment by richardTowers** [comment]
Just to add a bit to @tlwr's comment:

The cloud foundry team say that enabling bosh-dns globally (via runtime config) [is now a prerequisite](https://github.com/cloudfoundry/cf-deployment/releases/tag/v3.0.0).

If your concourse is deployed from the same bosh director as cloud foundry this issue will break subsequent CI jobs in difficult to debug ways. Specifically, the backup nameserver that @tlwr mentions made this even more tricky to diagnose.

Rather than getting DNS failures we see extremely slow operations (e.g. `cf auth`) while while waiting for the DNS clients to time out waiting for bosh-dns (falling back to the backup DNS).

As well as the eventual workaround @tlwr describes we first attempted to use the `networking-release` to set an `iptables` rule allowing access. It turns out this is not possible due to cloudfoundry/networking-release#15  - it will only let us append rules after the catch all REJECT on the INPUT chain.

This issue may be best fixed somewhere other than `concourse/concourse` (e.g. garden-runc-release?), but it seems likely that this issue will cause similar issues for other users attempting to upgrade cloudfoundry. It would be good to see a fix somewhere in the ecosystem or a workaround in concourse.

**Comment by rosenhouse** [comment]
Here are two ideas on how we could address this using garden-runc:

1. via legacy networker: enhance `guardian`'s [kawasaki networker](https://github.com/cloudfoundry/guardian/tree/master/kawasaki) so that it supports `INPUT` table rules, similar to [how the Silk CNI plugin does on cloudfoundry](https://github.com/cloudfoundry/silk-release/blob/b949495e16b6c3acab765e139408a5db57fa27c0/jobs/silk-cni/spec#L62-L67)

2. via CNI: configure guardian to use an external CNI plugin, perhaps via the [CF garden-external-networker](https://github.com/cloudfoundry/cf-networking-release/tree/develop/src/garden-external-networker) and then use a CNI plugin like [Silk](https://github.com/cloudfoundry/silk) that supports INPUT rules, or chain a basic CNI plugin (e.g. [bridge](https://github.com/containernetworking/plugins/tree/master/plugins/main/bridge) or [ptp](https://github.com/containernetworking/plugins/tree/master/plugins/main/ptp)) with a separate plugin capable of setting the INPUT rules (I don't think [portmap](https://github.com/containernetworking/plugins/tree/master/plugins/meta/portmap) supports that today).

**Comment by aminjam** [comment]
Diego wanted to report a persistent failure that was caused by not being able to resolve bosh-dns. We wanted to leave a comment in this feature request for anyone else that may experience the same failure.

We saw this failure happen after upgrading to bbl 7 (which uploads bosh-dns runtime-config for concourse as well as the CF). The failure was discovered when running [uptimer](https://code.cloudfoundry.org/uptimer). The failure manifested itself when running `cf api` and `cf auth` right after each other.

We were able to bypass this failure by explicitly setting the garden dns-servers to be an accessible address(e.g. `169.254.169.254` for GCP)
